File: topics/openacc/solutions/shared/blur_openacc.cpp
#include <iostream>
#include <cassert>

#include <omp.h>

#define NO_CUDA
#include "util.h"

#ifdef _OPENACC
#   pragma acc routine seq
#endif
double blur(int pos, const double *u)
{
    return 0.25*(u[pos-1] + 2.0*u[pos] + u[pos+1]);
}

void blur_twice_gpu_naive(double* in , double* out , int n, int nsteps)
{
    double* buffer = malloc_host<double>(n);

    for (auto istep = 0; istep < nsteps; ++istep) {
        int i;
        #pragma acc parallel loop copyin(in[0:n]) copyout(buffer[0:n])
        for(i=1; i<n-1; ++i) {
            buffer[i] = blur(i, in);
        }

        #pragma acc parallel loop copyin(buffer[0:n]) copy(out[0:n])
        for(i=2; i<n-2; ++i) {
            out[i] = blur(i, buffer);
        }

        // swap in/out
        std::swap(in, out);
    }

    free(buffer);
}

void blur_twice_gpu_nocopies(double *in , double *out , int n, int nsteps)
{
    double *buffer = malloc_host<double>(n);

    #pragma acc data copyin(in[0:n]) copy(out[0:n]) create(buffer[0:n])
    {
        for (auto istep = 0; istep < nsteps; ++istep) {
            int i;
            #pragma acc kernels
            {
                #pragma acc loop independent
                for(i = 1; i < n-1; ++i) {
                    buffer[i] = blur(i, in);
                }

                #pragma acc loop independent
                for(i = 2; i < n-2; ++i) {
                    out[i] = blur(i, buffer);
                }

                #pragma acc loop independent
                for (i = 0; i < n; ++i) {
                    in[i] = out[i];
                }
            }

        }
    }

    free(buffer);
}

int main(int argc, char** argv) {
    size_t pow    = read_arg(argc, argv, 1, 20);
    size_t nsteps = read_arg(argc, argv, 2, 100);
    size_t n = (1 << pow) + 4;

    auto size_in_bytes = n * sizeof(double);

    std::cout << "dispersion 1D test of length n = " << n
              << " : " << size_in_bytes/(1024.*1024.) << "MB"
              << std::endl;

    auto x0 = malloc_host<double>(n+4, 0.);
    auto x1 = malloc_host<double>(n+4, 0.);

    // set boundary conditions to 1
    x0[0]   = x1[0]   = 1.0;
    x0[1]   = x1[1]   = 1.0;
    x0[n-2] = x1[n-2] = 1.0;
    x0[n-1] = x1[n-1] = 1.0;

    auto tstart = get_time();
    blur_twice_gpu_nocopies(x0, x1, n, nsteps);
    auto time = get_time() - tstart;

    for(auto i=0; i<std::min(decltype(n){20},n); ++i) {
        std::cout << x1[i] << " ";
    }

    std::cout << "\n";
    std::cout << "==== that took " << time << " seconds"
              << " (" << time/nsteps << "s/step)" << std::endl;

    return 0;
}

File: topics/openacc/solutions/shared/blur_openacc.f90
real(kind(0d0)) function blur(pos, u, n)
  !$acc routine seq
  integer, intent(in) :: pos, n
  real(kind(0d0)), intent(in) :: u(n)

  blur = 0.25*(u(pos-1) + 2.0*u(pos) + u(pos+1))
end function blur

subroutine blur_twice_host(nsteps, n, in, out)
  use util
  implicit none
  integer, intent(in) :: n, nsteps
  real(kind(0d0)), intent(inout) :: in(n)
  real(kind(0d0)), intent(inout) :: out(n)

  integer istep, i
  real(kind(0d0)), dimension(:), allocatable :: buffer
  real(kind(0d0)), external :: blur
  !$acc routine(blur) seq

  allocate(buffer(n))

  do istep = 1,nsteps
     !$omp parallel do
     do i = 2,n-1
        buffer(i) = blur(i, in, n)
     enddo

     !$omp parallel do
     do i = 3,n-2
        out(i) = blur(i, buffer, n)
     enddo
     call swap(in, out)
  enddo

  deallocate(buffer)
end subroutine blur_twice_host

subroutine blur_twice_gpu_naive(nsteps, n, in, out)
  use util
  implicit none
  integer, intent(in) :: n, nsteps
  real(kind(0d0)), intent(inout) :: in(n)
  real(kind(0d0)), intent(inout) :: out(n)

  integer istep, i
  real(kind(0d0)), dimension(:), allocatable :: buffer
  real(kind(0d0)), external :: blur
  !$acc routine(blur) seq

  allocate(buffer(n))

  do istep = 1,nsteps
     !$acc parallel loop copyin(in) copyout(buffer)
     do i = 2,n-1
        buffer(i) = blur(i, in, n)
     enddo

     !$acc parallel loop copyin(buffer) copy(out)
     do i = 3,n-2
        out(i) = blur(i, buffer, n)
     enddo

     call swap(in, out)
  enddo

  deallocate(buffer)
end subroutine blur_twice_gpu_naive

subroutine blur_twice_gpu_nocopies(nsteps, n, in, out)
  implicit none
  integer, intent(in) :: n, nsteps
  real(kind(0d0)), intent(inout) :: in(n)
  real(kind(0d0)), intent(inout) :: out(n)

  integer istep, i
  real(kind(0d0)), dimension(:), allocatable :: buffer
  real(kind(0d0)), external :: blur
  !$acc routine(blur) seq

  allocate(buffer(n))

  !$acc data copyin(in) copy(out) create(buffer)
  do istep = 1,nsteps
     !$acc parallel
     !$acc loop independent
     do i = 2,n-1
        buffer(i) = blur(i, in, n)
     enddo

     !$acc loop independent
     do i = 3,n-2
        out(i) = blur(i, buffer, n)
     enddo

     !$acc loop independent
     do i = 1,n
        in(i) = out(i)
     enddo
     !$acc end parallel
  enddo
  !$acc end data

  deallocate(buffer)
end subroutine blur_twice_gpu_nocopies

program main
  use util
  implicit none

  integer pow, n, nsteps, err, i
  real(kind(0d0)), dimension(:), allocatable :: x0, x1
  real(kind(0d0)) :: blur_start, time_blur
  pow    = read_arg(1, 20)
  nsteps = read_arg(2, 100)
  n      = 2**pow + 4

  write(*, '(a i0 a f0.6 a)') 'dispersion 1D test of length n = ', n, ' : ', 8.*n/1024**2, 'MB'

  allocate(x0(n), x1(n), stat=err)
  if (err /= 0) then
     stop 'failed to allocate arrays'
  endif

  x0(1)   = 1.0
  x0(2)   = 1.0
  x0(n-1) = 1.0
  x0(n)   = 1.0

  x1(1)   = x0(1)
  x1(2)   = x0(2)
  x1(n-1) = x0(n-1)
  x1(n)   = x0(n)

  blur_start = get_time()
  call blur_twice_gpu_nocopies(nsteps, n, x0, x1)
  time_blur = get_time() - blur_start

  write(*, '(a f0.6 a f0.8 a)') '==== that took ', time_blur, &
       ' seconds (', time_blur/nsteps, 's/step )'
  do i = 1, min(20,n)
     write(*, '(f0.6 a)', advance='no') x1(i), ' '
  enddo
  write(*,*)

end program main
